packages/core-data/src/types/comment.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Internal dependencies
+ */
+import { AvatarUrls, Context, ContextualField, RenderedText } from './common';
+
+interface FullComment< C extends Context > {
+	/**
+	 * Unique identifier for the comment.
+	 */
+	id: number;
+	/**
+	 * The ID of the user object, if author was a user.
+	 */
+	author: number;
+	/**
+	 * Email address for the comment author.
+	 */
+	author_email: ContextualField< string, 'edit', C >;
+	/**
+	 * IP address for the comment author.
+	 */
+	author_ip: ContextualField< string, 'edit', C >;
+	/**
+	 * Display name for the comment author.
+	 */
+	author_name: string;
+	/**
+	 * URL for the comment author.
+	 */
+	author_url: string;
+	/**
+	 * User agent for the comment author.
+	 */
+	author_user_agent: ContextualField< string, 'edit', C >;
+	/**
+	 * The content for the comment.
+	 */
+	content: RenderedText< C >;
+	/**
+	 * The date the comment was published, in the site's timezone.
+	 */
+	date: string;
+	/**
+	 * The date the comment was published, as GMT.
+	 */
+	date_gmt: ContextualField< string, 'view' | 'edit', C >;
+	/**
+	 * URL to the comment.
+	 */
+	link: string;
+	/**
+	 * The ID for the parent of the comment.
+	 */
+	parent: number;
+	/**
+	 * The ID of the associated post object.
+	 */
+	post: ContextualField< number, 'view' | 'edit', C >;
+	/**
+	 * State of the comment.
+	 */
+	status: ContextualField< string, 'view' | 'edit', C >;
```

Comment:
I generated this from the REST API schema where it's declared as a string, same in the documentation. However, reading more into the controller implementation, I see there's a finite catalogue of these statuses, like `hold`, `approve`, `spam`, and `trash`:

https://github.com/WordPress/WordPress/blob/76207c890b68098000be3e54385dea54229a1944/wp-includes/comment.php#L2397

I also see, there's a possibility they're `1` or `0`, I'm not sure what to make of that. Intuitively it looks like a legacy way to denote a comment. I'll add a type here with all these possibilities, and if that turns out to be the wrong way to model it then someone will adjust it in the future.

Reply:
By the way, what's your take on union types vs enums? When would you use one instead of the other? e.g.

```ts
export type CommentStatus = 'hold' | 'approve' | 'spam' | 'trash' | '1' | '0';
```

vs

```ts
enum CommentStatus {
hold = 'hold';
approve = 'approve';
spam = 'spam';
trash = 'trash';
1 = '1';
0 = '0'; 
}
```

Reply:
I'm generally opposed to `enum` in TypeScript unless we have a compelling case for it. In almost call cases we have a strong prejudice for `const enum`. The difference is a lot of code overhead that multiplies its impact on small modules, and the organization we often have with `constants.ts` modules and the like. `const enum` are just values with no runtime object property lookup and no imports.

Here I'm fine with a string type union and I don't think it's going to end up mattering too much. One benefit of using an enum type is checking status:

```ts
if ( comment.status === CommentStatus.Approved ) {
```

but that comes with caveats, made considerably worse by the fact that we expose raw API types at all and don't parse them (such a tragedy we expose these; we should parse them), and that is that it's clear that we don't really have an enum in the traditional sense since multiple API values map to the same real status. further, in some cases I'm sure people would rather see the more concrete value so they can check it against the API data.

```ts
if ( comment.status === 'approve' ) {
```

herein lies an illustration of one of the bigger problems of exposing these external data types into the application where they don't belong: it's obvious we'll have a host of bugs related to people over-simplifying the comment status no matter how we represent it, because our representation is not of a "comment" but rather "the comment API."

So here again I think this leads us back to:
 - Wouldn't it be great if we never allowed this data into the application? (pipe dream, water under the bridge, you can't fight a tsunami)
 - Maybe err on the side of the descriptive type union since that evokes less sense of "this is one of these states" and more a sense of "its state could be described by one of these indicators."

Reply:
Awesome, thank you for elaborating! I published a spin-off PR to explore parsing API responses, let's move this part of the discussion there: https://github.com/WordPress/gutenberg/pull/38953